### PR TITLE
Add experimental command to write generated curation to documentation extension files

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://98528457

## Summary

This adds an experimental command to write the auto-generated curation to documentation extensions files. 

In some workflows outside of build integrations this can be a way to generate a starting point for curation and to determine what symbol disambiguation is needed to curate symbols with name collisions.

The new command also has two options to control which symbols to emit documentation extensions files for: `--from-symbol <symbol-link>` and `--depth <limit>`

```
OVERVIEW: Write documentation extension files into the DocC Catalog with one that supports a static hosting environment.

USAGE: docc process-catalog emit-generated-curation [<documentation-catalog>] [--additional-symbol-graph-dir <additional-symbol-graph-dir>] [--output-path <output-path>] [--from-symbol <symbol-link>] [--depth <limit>]

INPUTS & OUTPUTS:
  <documentation-catalog> Path to the DocC Catalog ('.docc') directory.
  --additional-symbol-graph-dir <additional-symbol-graph-dir>
                          Path to a directory of additional symbol graph files.
  --output-path <output-path>
                          The location where docc writes the transformed catalog.
        If no output-path is provided, docc will perform an in-place transformation of the provided DocC Catalog.

GENERATION OPTIONS:
  --from-symbol <symbol-link>
                          A link to a symbol to start generating documentation extension files from.
        If no symbol-link is provided, docc will generate documentation extension files starting from the module.
  --depth <limit>         A depth limit for which pages to generate documentation extension files for.
        If no depth is provided, docc will generate documentation extension files for all pages from the starting point.
        If 0 is provided, docc will generate documentation extension files for only the starting page.
        If a positive number is provided, docc will generate documentation extension files for the starting page and its descendants up to that depth limit (inclusive).

OPTIONS:
  -h, --help              Show help information.
```

For example, passing `--depth 0` will only emit a documentation extension file for the module (emitting the auto-generated curation for top-level symbols).
```
swift run docc process-catalog emit-generated-curation \
  YourProject.docc \
  --additional-symbol-graph-dir /path/to/symbol-graph-directory \
  --depth 0
```

Passing `--from-symbols SomeClass` will emit documentation extension files for "SomeClass" and its sub hierarchy.
```
swift run docc process-catalog emit-generated-curation \
  YourProject.docc \
  --additional-symbol-graph-dir /path/to/symbol-graph-directory \
  --from-symbol "SomeClass"
```

Passing `--from-symbols SomeClass` and `--depth 1` will emit documentation extension files for "SomeClass" and its direct descendants.
```
swift run docc process-catalog emit-generated-curation \
  YourProject.docc \
  --additional-symbol-graph-dir /path/to/symbol-graph-directory \
  --from-symbol "SomeClass" \
  --depth 0
```

Passing neither `--from-symbols` nor `--depth` will emit documentation extension files for every symbol in the project.
```
swift run docc process-catalog emit-generated-curation \
  YourProject.docc \
  --additional-symbol-graph-dir /path/to/symbol-graph-directory
```

If you want the generated documentation extension files to be emitted in a separate location you can pass an explicit output path.
```
swift run docc process-catalog emit-generated-curation \
  YourProject.docc \
  --additional-symbol-graph-dir /path/to/symbol-graph-directory \
  --output-path NewCatalog.docc
```

## Dependencies

None

### In a "process catalog" workflow

This functionality is mainly intended for workflows that aren't tied to a build integration and that can call a docc command other than "convert." Because of this, most of the configurability is only available in the new `process-catalog` command

- Emit generated curation for symbol graphs;
  ```
   swift run docc \
   process-catalog \
   emit-generated-curation \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory
  ```
  A new "Generated.docc" catalog will be created in the current directory
  
- Inspect the generated documentation extensions; `find Generated.docc`

- Manually recreate some of the generated curation. 
  - Open one of the generated documentation extension 
  - Create a few new topic sections and move some of the generated curation links into those topic sections
  - Remove the remaining generated curation links
  - Remove the other generated documentation extension files
  - Rename the generated DocC catalog
  
- Emit generated curation in addition to what's manually curated
  ```
   swift run docc \
   process-catalog \
   emit-generated-curation \
   YourRenamedCatalog.docc \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory
  ```
  
- Inspect the generated documentation extensions; `find YourRenamedCatalog.docc `
  - The removed documentation extensions files will be regenerated
  - Open the documentation extension file with manual curation
    - Symbols that are manually curated doesn't get generated curation links 

- Remove some manual curation and some generated documentation extension files again

- Add an article to the documentation catalog
  ```
  echo "# An article\n\nThis is an article" > YourRenamedCatalog.docc/article.md
  ```

- Emit generated curation into a different output location
  ```
   swift run docc \
   process-catalog \
   emit-generated-curation 
   YourRenamedCatalog.docc \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory \
   --output-path NewCatalog.docc
  ```
  
- Inspect the input documentation catalog; `find YourRenamedCatalog.docc`
  This catalog shouldn't be modified by the previous command
  
- Inspect the output documentation catalog; `find NewCatalog.docc`
  - The generated curation links and generated documentation extension files should exist in the catalog
  - The article and other non-generated files are not copied over from the input catalog to the output catalog.

- Emit only a documentation file for the module
  ```
   swift run docc \
   process-catalog \
   emit-generated-curation 
   YourRenamedCatalog.docc \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory \
   --depth 0 
  ```
  
- Inspect the documentation catalog; `find YourRenamedCatalog.docc`
  - Only one generated documentation extension file should exist in the catalog (for the module)

- Emit documentation files for the specific symbol sub hierarchy
  ```
   swift run docc \
   process-catalog \
   emit-generated-curation 
   YourRenamedCatalog.docc \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory \
   --from-symbol SomeClass
  ```
  
- Inspect the documentation catalog; `find YourRenamedCatalog.docc`
  - Only generated documentation extension files for "SomeClass" and its descendants should have been added.

- Try emitting documentation files for the symbol that doesn't exist
  ```
   swift run docc \
   process-catalog \
   emit-generated-curation 
   YourRenamedCatalog.docc \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory \
   --from-symbol SomeClas
  ```
  
  - There should be a warning that "SomeClas" wasn't found. If the symbol path is similar to a real symbol there may be suggestions on how to fix the symbol link.
  
### In a "convert" workflow

Add the `--experimental-modify-catalog-with-generated-curation` flag to a `docc convert` call that uses a documentation catalog,

After the conversion, the catalog may contain generated documentation extension files with auto-generated curation. 

- Create an empty catalog; `mkdir Empty.docc`
- Convert some symbol graphs; 
   ```
   swift run docc convert \
   Empty.docc \
   --experimental-modify-catalog-with-generated-curation \
   --additional-symbol-graph-dir /path/to/symbol-graph-directory
   ```
- Check the content of the catalog; `find Empty.docc`

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
Yes, without IDE integration there isn't a great way to get to this information today. I'm not sure if this is the _best_ way to surface this information but I think it's a decent start that we can evolve as more people use it and provide feedback on it.

Inspecting the linkable entities is a decent workaround but it's has some drawbacks since the path to a page has more restrictions than the link to a page. For example, `<(_:_:)` and `>(_:_:)` are valid _links_ but they're not valid paths, so the `<`, and `>` characters are replaced with `_` in the path. This makes both paths `_(_:_:)` which introduce the need for disambiguation. Since both symbols are operators, the only other way to distinguish them is by adding the hash of their unique identifiers. This means that the path you'll find in the linkable entities would be something like `_(_:_:)-abc123` and `_(_:_:)-def456` which aren't the best _links_ to those pages.

**Comment by heckj** [comment]
@d-ronnqvist I wanted to give your new CLI a shot, but I think there's a missing file from this PR - I'm getting compiler errors on attempting to build - both this branch explicitly, and rebased over `apple/main` (just in case). The type `ProcessCatalog` isn't resolving.

Error example:
```
Building for debugging...
/Users/heckj/src/swift-project/swift-docc/Sources/SwiftDocCUtilities/Docc.swift:14:133: error: cannot find 'ProcessCatalog' in scope
private let subcommands: [ParsableCommand.Type] = [Docc.Convert.self, Docc.Index.self, Docc.Preview.self, Docc.ProcessArchive.self, ProcessCatalog.self]
                                                                                                                                    ^~~~~~~~~~~~~~
/Users/heckj/src/swift-project/swift-docc/Sources/SwiftDocCUtilities/Docc.swift:14:133: error: cannot find 'ProcessCatalog' in scope
private let subcommands: [ParsableCommand.Type] = [Docc.Convert.self, Docc.Index.self, Docc.Preview.self, Docc.ProcessArchive.self, ProcessCatalog.self]
                                                                                                                                    ^~~~~~~~~~~~~~
/Users/heckj/src/swift-project/swift-docc/Sources/SwiftDocCUtilities/Docc.swift:14:133: error: cannot find 'ProcessCatalog' in scope
private let subcommands: [ParsableCommand.Type] = [Docc.Convert.self, Docc.Index.self, Docc.Preview.self, Docc.ProcessArchive.self, ProcessCatalog.self]
                                                                                                                                    ^~~~~~~~~~~~~~
```

**Comment by d-ronnqvist** [comment]
Apologies. I think I had missed that the other commands were prefixed with `Docc.` when I resolved a merge conflict. That should be fixed now.

**Comment by boisy** [comment]
@d-ronnqvist, a minor, minor nit... there's an extra space between "with" and "one" on line 20 of EmitGeneratedCuration.swift.

**Comment by d-ronnqvist** [comment]
> @d-ronnqvist, a minor, minor nit... there's an extra space between "with" and "one" on line 20 of EmitGeneratedCuration.swift.

Oh. That sentence is also a mix of two different abstracts. I should go through all the CLI text before marking this ready to review.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/EmitGeneratedCuration.swift [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import Foundation
+import ArgumentParser
+
+extension Docc.ProcessCatalog {
+    /// Emits documentation extension files that reflect the auto-generated curation.
+    struct EmitGeneratedCuration: ParsableCommand {
+        
+        static var configuration = CommandConfiguration(
+            commandName: "emit-generated-curation",
+            abstract: "Write documentation extension files into the DocC Catalog with one that supports a static hosting environment.")
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/EmitGeneratedCuration.swift:20]: No. I don't know where that came from. I must have made a copy&paste mistake.

[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/EmitGeneratedCuration.swift]: Fixed in https://github.com/apple/swift-docc/pull/597/commits/deb72d446ca8fd8cd6593e18e0f75e66f31cfcaf

File: Package.resolved [hunk 2]
```diff
@@ -86,8 +86,8 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/apple/swift-nio.git",
       "state" : {
-        "revision" : "2d8e6ca36fe3e8ed74b0883f593757a45463c34d",
```
Comment (daniel-grumberg) [inline Package.resolved:89]: Are we expecting this update?

File: Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        var text = ""
+        for taskGroup in automaticTopics {
+            if isForTopLevelCuration, let firstReference = taskGroup.references.first, context.documentationCache[firstReference]?.symbol?.kind.identifier == .typeProperty {
+                // Skip type properties in top-level curation. It's not clear what's the right place for these symbols are since they exist in
+                // different places in different source languages (which documentation extensions don't yet have a way of representing).
+                continue
+            }
+            
+            let links: [(link: String, comment: String?)] = taskGroup.references.compactMap { (curatedReference: ResolvedTopicReference) -> (String, String?)? in
+                guard !alreadyCurated.contains(curatedReference.absoluteString) else { return nil }
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift:65]: IIUC this only checks if the reference we are looking at was curated as a direct descendent of the node we are currently processing?

File: Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+            }
+            
+            guard !links.isEmpty else { continue }
+            
+            text.append("\n\n### \(taskGroup.title ?? "<!-- This auto-generated topic has no title -->")\n")
+            
+            let longestLink = links.map(\.link.count).max()! /* `links` are non-empty */ + 7 /* the extra characters used to make the link "\n- ``" before and "``" after */
+            for (link, comment) in links {
+                if let comment = comment {
+                    text.append("\n- ``\(link)``".padding(toLength: longestLink, withPad: " ", startingAt: 0))
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift:86]: Is there something I am missing here, but why would whitespace padding here be useful?

File: Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                }
+                return (linkInfo.link, nil)
+                
+            }
+            
+            guard !links.isEmpty else { continue }
+            
+            text.append("\n\n### \(taskGroup.title ?? "<!-- This auto-generated topic has no title -->")\n")
+            
+            let longestLink = links.map(\.link.count).max()! /* `links` are non-empty */ + 7 /* the extra characters used to make the link "\n- ``" before and "``" after */
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift:83]: This is a slightly confusing variable name, took a little while to understand this measures the longest line without a comment. I wonder if generating the String components first and then calculating padding might be more readable, even though it introduces an extra loop over the links, something along these lines:
```swift
let linkTextFragments = links.map { (link: "\n- ``\($0.link)``", comment: $0.comment.map { " <!-- \($0) -->" }) }
let longestLinkLine = linkTextFragments.map(\.link.count).max()!
let curationText = linkTextFragments
    .map { $0.comment == nil ? $0.link : $0.link.padding(toLength: longestLinkLine, withPath: " ", startingAt: 0).append($0.comment!) }
    .joined()
text.append(curationText)
```

**Review by daniel-grumberg — APPROVED** [review]

File: Package.resolved (hunk 2 above)
Comment (d-ronnqvist) [reply Package.resolved:89]: No. I'll fix it when I merge main.

File: Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift (hunk 3 above)
Comment (daniel-grumberg) [reply Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift:65]: Fantastic! I think that will simplify the code somewhat.

File: Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift (hunk 5 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Catalog Processing/GeneratedCurationWriter.swift:83]: I incorporated the code that wraps the link and comment text in markup into the `compactMap` above. That made the code that computes the longest link and appends the markup strings easier to follow.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Review by daniel-grumberg — APPROVED** [review]

**Pull request merged**